kernel/cpu_local/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+//! Offers types and macros to declare and access CPU-local storage (per-CPU variables).
+//!
+//! CPU-local variables cannot be used until after a given CPU has been initialized,
+//! i.e., its Local APIC (on x86_64) has been discovered and properly configured.
+//! Currently, the [`init()`] routine in this crate should be invoked by
+//! another init routine from the `per_cpu` crate.
+//!
+//! Note that Rust offers the `#[thread_local]` attribute for thread-local storage (TLS),
+//! but there is no equivalent for CPU-local storage.
+//! On x86_64, TLS areas use the `fs` segment register for the TLS base,
+//! and this crates uses the `gs` segment register for the CPU-local base.
+
+#![no_std]
+#![feature(thread_local)]
+extern crate alloc;
+
+use core::{
+    arch::asm,
+    marker::PhantomData,
+    mem::{size_of, align_of},
+};
+use alloc::collections::{BTreeMap, btree_map::Entry};
+use memory::{MappedPages, PteFlags};
+use preemption::{hold_preemption, PreemptionGuard};
+use spin::Mutex;
+
+#[cfg(target_arch = "x86_64")]
+use x86_64::{registers::model_specific::GsBase, VirtAddr};
+
+/// Info use to obtain a reference to a CPU-local variable; see [`CpuLocal::new()`].
+pub struct FixedCpuLocal {
+    offset: usize,
+    size:   usize,
+    align:  usize,
+}
+// NOTE: These fields must be kept in sync with `cpu_local::FixedCpuLocal`.
+impl FixedCpuLocal {
+    const SELF_PTR_OFFSET: usize = 0;
+    pub const CPU_ID:                       Self = Self { offset: 8,  size: 4, align: 4 };
+    pub const PREEMPTION_COUNT:             Self = Self { offset: 12, size: 1, align: 1 };
+    pub const TASK_SWITCH_PREEMPTION_GUARD: Self = Self { offset: 16, size: 8, align: 4 };
+    pub const DROP_AFTER_TASK_SWITCH:       Self = Self { offset: 24, size: 8, align: 8 };
+}
+
+/// A reference to a CPU-local variable.
+///
+/// Note that this struct doesn't contain an instance of the type `T`,
+/// and dropping it has no effect.
+pub struct CpuLocal<const OFFSET: usize, T>(PhantomData<*mut T>);
+impl<const OFFSET: usize, T> CpuLocal<OFFSET, T> {
+    /// Creates a new reference to a predefined CPU-local variable.
+    ///
+    /// # Arguments
+    /// * `fixed`: information about this CPU-local variable.
+    ///    Must be one of the public constants defined in [`FixedCpuLocal`].
+    ///
+    /// # Safety
+    /// This is unsafe because we currently do not have a way to guarantee
+    /// that the caller-provided type `T` is the same as the type intended for use
+    /// by the given `FixedCpuLocal`.
+    ///
+    /// Thus, the caller must guarantee that the type `T` is correct for the
+    /// given `FixedCpuLocal`.
+    pub const unsafe fn new_fixed(
```

Comment:
we don't use it yet, but it would be used like so:
```rust
// in `kernel/preemption`:
static PREEMPTION_COUNT: CpuLocal<..., PreemptionCount> = CpuLocal::new_fixed(FixedCpuLocal::PREEMPTION_COUNT);

// in hold_preemption, increase the count:
let prev_val = PREEMPTION_COUNT.with(|count| count.fetch_add(1, ...));

// in PreemptionGuard::drop(), decrease the count:
let prev_val = PREEMPTION_COUNT.with(|count| count.fetch_sub(1, ...));
```